File: gateway/kotlin/gateway/src/test/kotlin/dev/bridgething/gateway/MockAdapter.kt
package dev.bridgething.gateway

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.consumeAsFlow

/**
 * Test-only [Adapter] that lets a test drive byte/connection events into the
 * gateway and pull back outbound frames. The events flow is backed by an
 * unlimited Channel so simulate() before the gateway has had a chance to
 * subscribe still gets buffered — a SharedFlow with no replay drops events
 * for a not-yet-attached subscriber, which is exactly the race that bit the
 * forwards-events test on the first run.
 */
internal class MockAdapter : Adapter {
  private val incomingEvents: Channel<AdapterEvent> = Channel(Channel.UNLIMITED)
  override val events: Flow<AdapterEvent> = incomingEvents.consumeAsFlow()

  val sentFrames: Channel<Pair<String, ByteArray>> = Channel(Channel.UNLIMITED)

  var startCalled: Boolean = false
  var stopCalled: Boolean = false
  val disconnectCalls: MutableList<String> = mutableListOf()

  override suspend fun start() { startCalled = true }
  override suspend fun stop() {
    stopCalled = true
    incomingEvents.close()
    sentFrames.close()
  }
  override suspend fun disconnect(deviceId: String) {
    disconnectCalls.add(deviceId)
  }
  override suspend fun send(deviceId: String, frame: ByteArray) {
    sentFrames.send(deviceId to frame)
  }

  suspend fun simulate(event: AdapterEvent) {
    incomingEvents.send(event)
  }
}

File: packages/gateway/swift/Tests/BridgethingGatewayTests/MockAdapter.swift
import Foundation
@testable import BridgethingGateway

/// Test-only `Adapter` that lets a test drive byte/connection events into the
/// gateway and pull back outbound frames. Not thread-safe beyond what
/// `AsyncStream`'s continuations already guarantee, which is sufficient for
/// single-test usage.
final class MockAdapter: Adapter, @unchecked Sendable {
  let events: AsyncStream<AdapterEvent>
  private let eventContinuation: AsyncStream<AdapterEvent>.Continuation

  let sentFrames: AsyncStream<(deviceId: String, frame: Data)>
  private let sentContinuation: AsyncStream<(deviceId: String, frame: Data)>.Continuation

  var startCalled = false
  var stopCalled = false
  var disconnectCalls: [String] = []

  init() {
    let (e, ec) = AsyncStream.makeStream(of: AdapterEvent.self)
    let (s, sc) = AsyncStream.makeStream(of: (deviceId: String, frame: Data).self)
    events = e
    eventContinuation = ec
    sentFrames = s
    sentContinuation = sc
  }

  func start() async throws { startCalled = true }
  func stop() async {
    stopCalled = true
    eventContinuation.finish()
    sentContinuation.finish()
  }

  func disconnect(deviceId: String) async throws {
    disconnectCalls.append(deviceId)
  }

  func send(deviceId: String, frame: Data) async throws {
    sentContinuation.yield((deviceId: deviceId, frame: frame))
  }

  func simulate(_ event: AdapterEvent) {
    eventContinuation.yield(event)
  }
}
